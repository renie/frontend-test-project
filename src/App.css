.App {
  margin: 0 auto;
  max-width: 1200px;
}

.App h1,
.App code {
  display: block;
  margin: 20px auto 40px;
  text-align: center;
}

.List {
  text-align: left;
}
