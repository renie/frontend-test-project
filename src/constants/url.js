export const getAPIUrl = endpoint => `https://jsonplaceholder.typicode.com/${endpoint}`;
