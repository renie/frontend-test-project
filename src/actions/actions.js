import { getAPIUrl } from '../constants/url.js'
import {
  GET_USERS_REQUEST,
  GET_USERS_SUCCESS,
  GET_USERS_ERROR
} from '../constants/actionTypes.js'


const requestUsers = () => ({ type: GET_USERS_REQUEST });
const receiveUsers = users => ({ type: GET_USERS_SUCCESS, users });
const errorUsers = error => ({ type: GET_USERS_ERROR, error });

export const fetchUsers = () => dispatch => {
  dispatch(requestUsers());

  return fetch(getAPIUrl('users'))
      .then(
        response => response.json(),
        error => dispatch(errorUsers(error))
      )
      .then(
        json => dispatch(receiveUsers(json))
      )
      .catch(
        error => dispatch(errorUsers(error)))
}
