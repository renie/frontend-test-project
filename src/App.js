import React, { Component } from 'react'
import { connect } from 'react-redux'

import './App.css'
import CardList from './components/card/CardList'
import { fetchUsers } from './actions/actions'

export class App extends Component {
  constructor(props){
    super(props)
    this.state = {users: []}
    this.toggleCardBinded = this.toggleCard.bind(this)
  }

  addHiddenPropertyToUsers(users) {
    return users.map(user => ({...user, isHidden: false}))
  }

  toggleCard(userId) {
    this.setState({
      users: this.state.users.map(user => {
        if (user.id !== userId)
          return user

        const newUser = {...user}
        newUser.isHidden = !newUser.isHidden
        return newUser
      })
    })
  }

  componentDidMount() {
    this.props
      .fetchUsers()
      .then(action => {
        this.setState({
          users: this.addHiddenPropertyToUsers(action.users)
        })
      })
  }

  render() {
    return (
      <div className="App">
        <h1>Welcome to your Dashboard!</h1>
        <code>Good luck :)</code>
        <CardList users={this.state.users} toggleCard={this.toggleCardBinded}/>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  users: state.appReducer.app.users,
  lastRefresh: state.appReducer.app.lastRefresh
})

const mapDispatchToProps = dispatch => ({
  fetchUsers: () => dispatch(fetchUsers())
})

export default connect(mapStateToProps, mapDispatchToProps)(App)
