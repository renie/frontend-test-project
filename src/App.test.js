import React from 'react';
import ReactDOM from 'react-dom';
import { App } from './App';
import { shallow } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import fetchMock from 'fetch-mock';
import { getAPIUrl } from './constants/url.js';
import { fetchUsers } from './actions/actions';


const mockStore = configureMockStore([thunk])

describe('APP', () => {
  afterEach(() => {
    fetchMock.restore()
  });

  it('renders without crashing', () => {
    const expectedBody = [];
    const store = mockStore({ app: { users: [] }})
    const mockfetchUsers = () => store.dispatch(fetchUsers())

    fetchMock.getOnce(getAPIUrl('users'), {
      body: expectedBody,
      headers: { 'content-type': 'application/json' }
    })

    const wrapper = shallow(<App store={store} fetchUsers={mockfetchUsers} />);

    expect(wrapper).toMatchSnapshot();
  });
})
