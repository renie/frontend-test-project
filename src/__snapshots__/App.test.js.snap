// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`APP renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App
    fetchUsers={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1>
          Welcome to your Dashboard!
        </h1>,
        <code>
          Good luck :)
        </code>,
        <CardList
          toggleCard={[Function]}
          users={Array []}
        />,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Welcome to your Dashboard!",
        },
        "ref": null,
        "rendered": "Welcome to your Dashboard!",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Good luck :)",
        },
        "ref": null,
        "rendered": "Good luck :)",
        "type": "code",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "toggleCard": [Function],
          "users": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            Welcome to your Dashboard!
          </h1>,
          <code>
            Good luck :)
          </code>,
          <CardList
            toggleCard={[Function]}
            users={Array []}
          />,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Welcome to your Dashboard!",
          },
          "ref": null,
          "rendered": "Welcome to your Dashboard!",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Good luck :)",
          },
          "ref": null,
          "rendered": "Good luck :)",
          "type": "code",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "toggleCard": [Function],
            "users": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
